### PR TITLE
Use `ParamSpec` in a few places

### PR DESCRIPTION
**Pull request by DMRobertson**
I tried to concentrate on publicly visible functions which pass a `*args,  **kwargs` pair, so that we get some checking that we end up passing the right arguments.

Related: #11711 
Will probably conflict with #12666.

### CODE REVIEW

File: synapse/storage/databases/main/events.py [hunk 1]
```diff
@@ -1648,9 +1648,15 @@ def prefill():
...
-        txn.call_after(self.store._invalidate_get_event_cache, event.redacts)
+        """Invalidate the caches for the redacted event.
+
+        Note that these caches are also cleared as part of event replication in
+        _invalidate_caches_for_event.
+        """
+
+        # type-ignore: mypy detects that event.redacts may be None. Presumably the
+        # application has ensured that this is not the case if we call this function.
+        txn.call_after(self.store._invalidate_get_event_cache, event.redacts)  # type: ignore[arg-type]
```
Comment (erikjohnston) [inline synapse/storage/databases/main/events.py:1659]: I think I'd prefer an `assert events.redacts is not None`? We do check that before we call this function, so it shouldn't be a risky thing to do

**Review by erikjohnston — APPROVED** [review]

**Comment by DMRobertson** [comment]
olddeps doesn't like a class that's generic over a ParamSpec.

**Comment by DMRobertson** [comment]
> olddeps doesn't like a class that's generic over a ParamSpec.

Reproduce with `typing-extensions==3.10.0` on Python 3.8. Looks like we need https://github.com/python/typing/pull/817, in 3.10.0.1.

**Comment by DMRobertson** [comment]
@erikjohnston can you take another look at the followup changes? (Assuming CI passes now)

**Review by erikjohnston — APPROVED** [review]

**Pull request merged**